engines/wintermute/base/gfx/opengl/base_render_opengl3d.cpp

```diff
@@ -0,0 +1,369 @@
+bool Wintermute::BaseRenderOpenGL3D::setup2D(bool force) {
+	if (_state3D || force) {
+		_state3D = false;
+
+		// some states are still missing here
+
+		glDisable(GL_LIGHTING);
+		glDisable(GL_DEPTH_TEST);
+		glDepthMask(false);
+		glDisable(GL_STENCIL);
+		glDisable(GL_CLIP_PLANE0);
+		glDisable(GL_FOG);
+		glLightModeli(GL_LIGHT_MODEL_AMBIENT, 0);
+
+		glEnable(GL_CULL_FACE);
+		glCullFace(GL_CCW);
+		glEnable(GL_ALPHA_TEST);
+		glEnable(GL_BLEND);
+		glAlphaFunc(GL_GEQUAL, 0x00);
+		glPolygonMode(GL_FRONT, GL_FILL);
+		glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+
+		glActiveTexture(GL_TEXTURE0);
+		glTexEnvf(GL_TEXTURE_ENV, GL_COMBINE_RGB, GL_MODULATE);
+		glTexEnvf(GL_TEXTURE_ENV, GL_SRC0_RGB, GL_TEXTURE);
+		glTexEnvf(GL_TEXTURE_ENV, GL_SRC1_RGB, GL_PREVIOUS);
+		glTexEnvf(GL_TEXTURE_ENV, GL_COMBINE_ALPHA, GL_MODULATE);
+		glTexEnvf(GL_TEXTURE_ENV, GL_SRC0_ALPHA, GL_TEXTURE);
+		glTexEnvf(GL_TEXTURE_ENV, GL_SRC1_ALPHA, GL_PREVIOUS);
+
+		glActiveTexture(GL_TEXTURE1);
+		glDisable(GL_TEXTURE_2D);
+
+		glActiveTexture(GL_TEXTURE0);
+
+		setProjection2D();
+	}
+
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::setup3D(bool force) {
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::setupLines() {
+	return true;
+}
+
+Wintermute::BaseSurface* Wintermute::BaseRenderOpenGL3D::createSurface() {
+	return new BaseSurfaceOpenGL3D(nullptr, this);
+}
+
+void Wintermute::BaseRenderOpenGL3D::endSaveLoad() {
+
+}
+
+bool Wintermute::BaseRenderOpenGL3D::drawSprite(const OpenGL::Texture& tex, const Wintermute::Rect32& rect,
+												float zoomX, float zoomY, const Wintermute::Vector2& pos,
+												uint32 color, bool alphaDisable, Graphics::TSpriteBlendMode blendMode,
+												bool mirrorX, bool mirrorY) {
+	Vector2 scale(zoomX / 100.0f, zoomY / 100.0f);
+	return drawSpriteEx(tex, rect, pos, Vector2(0.0f, 0.0f), scale, 0.0f, color, alphaDisable, blendMode, mirrorX, mirrorY);
+}
+
+bool Wintermute::BaseRenderOpenGL3D::drawSpriteEx(const OpenGL::Texture& tex, const Wintermute::Rect32& rect,
+												  const Wintermute::Vector2& pos, const Wintermute::Vector2& rot, const Wintermute::Vector2& scale,
+												  float angle, uint32 color, bool alphaDisable, Graphics::TSpriteBlendMode blendMode,
+												  bool mirrorX, bool mirrorY) {
+	// original wme has a batch mode for sprites, we ignore this for the moment
+
+	// The ShaderSurfaceRenderer sets an array buffer which appearently conflicts with us
+	// Reset it!
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
+	float width = (rect.right - rect.left) * scale.x;
+	float height = (rect.bottom - rect.top) * scale.y;
+
+	float tex_left = (float) rect.left / (float) tex.getWidth();
+	float tex_top = (float) rect.top / (float) tex.getHeight();
+	float tex_right = (float) rect.right / (float) tex.getWidth();
+	float tex_bottom = (float) rect.bottom / (float) tex.getHeight();
+
+	float offset = _viewportRect.height() / 2.0f;
+	float corrected_y =  (pos.y - offset) * -1.0f + offset;
```

Comment:
additional space not necessary after =
there is not aligned to other line. so one space

Reply:
not done yet